src/amperity/gocd/secret/vault/plugin.clj

```diff
@@ -110,25 +128,83 @@
 ;; configuring a secret backend in GoCD.
 (defmethod handle-request "go.cd.secrets.get-metadata"
   [_ _ _]
-  ;; TODO: how does the plugin authenticate?
   {:response-code    200
    :response-headers {}
-   :response-body    [{:key      :vault_addr
-                       :metadata {:required true :secure false}}]})
+   :response-body    (mapv (fn [input-key]
+                             {:key      input-key
+                              :metadata (-> input-schema input-key :metadata)})
+                           (keys input-schema))})
+
+
+(defn- input-error-message
+  "If the input field is valid, does nothing, if it's not, returns a string describing the error.
+
+  Params:
+  - `field-key`: the input field you are validating
+  - `field-value`: the inputted value which you wish to verify
+  - `field-label`: the human readable representation of the field key"
+  [field-key field-value]
+  (let [field-model (field-key input-schema)
+        validate-func-errors (remove #((:func %)  field-value)
+                                     (:validate-funcs field-model))]
+    (cond
+      ;; field is required but empty
+      (and (-> field-model :metadata :required) (str/blank? field-value))
+      (str (:label field-model) " is required")
+
+      ;; field is not empty and there is an error
+      (not (or (str/blank? field-value) (empty? validate-func-errors)))
+      (:message (first validate-func-errors))
+
+      :else
+      nil)))
+
+
+(defn- authenticate-client-from-inputs!
+  "Authenticates the Vault Client
+
+  Params:
+  - `client` The Vault Client you wish to authenticate
+  - `inputs` A map containing the user inputted settings for the plugin"
+  [client inputs]
+  (case (keyword (:auth_method inputs))
+    :token
+    (vault/authenticate! client :token (:vault_token inputs))
+
+    (throw (ex-info "Could not recognize user inputted vault auth type"
```

Comment:
Maybe say "unhandled" instead of saying can't recognize. To me unhandled says "i see what you said, but don't support it", whereas the other is just "what?".